### ISSUE TITLE
Questions about polars cutqcut

### ISSUE THREAD

**Issue by lidh15** [pola-rs/polars#10525]
Maybe I'm confused but shouldn't this `blen` be `qbreaks.len()` rather than `ll.len()`?
```rust
                polars_ensure!(ll.len() == qbreaks.len() + 1,
                    ShapeMismatch: "Wrong number of labels");
                let blen = ll.len();
                Some(
                    ll.into_iter()
                        .enumerate()
                        .filter(|(i, _)| *i == 0 || *i == blen || qbreaks[*i] != qbreaks[i - 1])
                        .unzip::<_, _, Vec<_>, Vec<_>>()
                        .1,
                )
```
we enumerate `ll` and it seems that the index `i` won't reach `ll.len()` at all. 
however, here it should be checked that if `i` is greater than 0 and less than `qbreaks.len()` as we index `i` and `i-1` from qbreaks.
so I guess `let blen = ll.len();` should be `let blen = qbreaks.len()`?
and the varname `blen` is also reasonably short  for `breaks length` rather than `labels length`.
the code is at
[https://github.com/pola-rs/polars/blob/8a7ac2a490f1fac4fed3f0233b0494ff847af65d/crates/polars-ops/src/series/ops/cut.rs#L124-L133](L124-L133)

**Comment by orlp** [comment]
The looks suspicious indeed. We should always keep the first label, and after that the *last* label per group of equivalent breakpoints thereafter, whereas this keeps the first.

**Comment by orlp** [comment]
I think you are right that `blen` should be `blen - 1`.

**Comment by magarick** [comment]
This code is only ever used if there are empty bins (i.e. two consecutive breaks that are the same). It removes labels corresponding to empty intervals. However, you need at least two labels, since you have at least one valid break (unless all values are the same). So we take the first, the last, and labels not corresponding to 2 identical breaks in order. On the very next line, `qbreaks` is deduplicated as well.
This preserves the requirement that the number of labels is 1 more than the number of breaks (because n breaks means n+1 intervals)

**Comment by orlp** [comment]
@magarick Except `*i == blen` can never be true. `enumerate` uses zero-based indexing so the last element has index `blen - 1`.

**Comment by magarick** [comment]
Yeah, you're right. I'm surprised this worked at all. Easy fix though.

**Comment by lidh15** [comment]
> Yeah, you're right. I'm surprised this worked at all. Easy fix though.

hope to see the fix soon
I've locally cargo patched it, will it come with the next polars patch version?
